site-src/geps/gep-1651.md

```diff
@@ -0,0 +1,108 @@
+# GEP-1651: Cluster Local Gateways
+
+* Issue: [#1651](https://github.com/kubernetes-sigs/gateway-api/issues/1651)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Allow users to configure a Gateway so that it is only accessible within a 
+single cluster. We call this a cluster local Gateway.
+
+## Goals
+
+- Cluster local Gateways can be created in a declarative way
+- Addresses used by the cluster local gateway are only accessible within a 
+  single cluster
+
+## Non-Goals
+
+- Determining 'visibility' on a per-request basis
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an 
+application using Knative's HTTP routing support, but make it only available 
+within the cluster. I want to be able to specify both the "internal" 
+(service.namespace.svc) and "external" (service.namespace.example.com).
+Gateways using the same GatewayClass on the cluster, but ensure that the 
+"internal" service is only reachable within the cluster. This would greatly 
+simplify deployment for users over the instructions we have today.
+
+The reason why we require an "internal" Gateway is so we can take advantage of 
+layer 7 load balancing for our internal services.
+
+## API
```

Comment:
since this design proposes inserting a field under the already existing `addresses` field, I prefer if `Private` and `Public` is used instead of `ClusterIP` and `Loadbalancer`. Implementations can infer that to build ClusterIP or Loadbalancer type services

Reply:
Although I was originally thinking we might just want one term to cover "Internal", I think it may be more nuanced than that. In some cases, you can use annotations to configure a Service type=LB to be internal, for example, `cloud.google.com/load-balancer-type: "Internal"` or `service.beta.kubernetes.io/aws-load-balancer-scheme: "internal"`. That would mean that in-cluster implementations of the API would be able to represent both SameNetwork and SameCluster if there's sufficient demand.

Reply:
I think an internal LB (LB service with private IP) could be a valid implementation of `Private` .
I understand that it would remove the fine grained `SameCluster` option, but imho prefer starting with `Private` and `Public` and later creating more fine grained options like `PrivateNetwork` and `PrivateCluster` based on user feedback

Reply:
I agree that these are very similar, but since it's possible for some providers to offer both, we probably need to be careful here. Are there any in-cluster implementations that would like to be able to offer both? It seems like it would make a fairly significant difference for the user whether an ILB or ClusterIP Service is created. I'd hate to end up with a world where we had `Private`, `PrivateNetwork`, and `PrivateCluster` options here.

Reply:
I would be very careful about whether or not these terms are well-defined.

Example: private IPs -- some users include carrier grade NAT ranges as private. Others use public ranges they own.

Generally, if you refer to Kubernetes concepts, you are on solid ground. To go beyond that may be hard to generalize.